cirq-google/cirq_google/engine/stream_manager.py

```diff
@@ -134,27 +141,35 @@ def submit(
         Raises:
             ProgramAlreadyExistsError if the program already exists.
             StreamError if there is a non-retryable error while executing the job.
+            ValueError if program name is not set.
             concurrent.futures.CancelledError if the stream is stopped while a job is in flight.
             google.api_core.exceptions.GoogleAPICallError if the stream breaks with a non-retryable
                 error.
         """
+        if 'name' not in program:
+            raise ValueError('Program name must be set.')
+
         if self._manage_stream_loop_future is None:
             self._manage_stream_loop_future = self._executor.submit(self._manage_stream)
             self._manage_stream_loop_future.add_done_callback(self._manage_stream_cancel())
         return self._executor.submit(self._manage_execution, project_name, program, job)
 
     def stop(self) -> None:
         """Closes the open stream and resets all management resources."""
-        if self._manage_stream_loop_future is not None:
-            self._manage_stream_loop_future.cancel()
-        self._reset()
+        try:
+            if (
+                self._manage_stream_loop_future is not None
+                and not self._manage_stream_loop_future.done()
+            ):
+                self._manage_stream_loop_future.cancel()
+        finally:
+            self._reset()
 
     def _reset(self):
```

Comment:
Should "reset" cancel actively running tasks (call `stop()`)? That would cause a loop right now, but it makes sense that we'd clear old work as part of a reset.

Reply:
Chatted offline - this no longer applies